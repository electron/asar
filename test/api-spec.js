'use strict'

const assert = require('assert')
const fs = require('../lib/wrapped-fs')
const os = require('os')
const path = require('path')
const rimraf = require('rimraf')

const asar = require('..')
const compDirs = require('./util/compareDirectories')
const compFileLists = require('./util/compareFileLists')
const compFiles = require('./util/compareFiles')
const transform = require('./util/transformStream')

async function assertPackageListEquals (actualList, expectedFilename) {
  const expected = await fs.readFile(expectedFilename, 'utf8')
  return compFileLists(actualList.join('\n'), expected)
}

describe('api', function () {
  beforeEach(() => { rimraf.sync(path.join(__dirname, '..', 'tmp'), fs) })

  it('should create archive from directory', async () => {
    await asar.createPackage('test/input/packthis/', 'tmp/packthis-api.asar')
    return compFiles('tmp/packthis-api.asar', 'test/expected/packthis.asar')
  })
  if (os.platform() === 'win32') {
    it('should create archive with windows-style path separators', async () => {
      await asar.createPackage('test\\input\\packthis\\', 'tmp\\packthis-api.asar')
      return compFiles('tmp/packthis-api.asar', 'test/expected/packthis.asar')
    })
  }
  it('should create archive from directory (without hidden files)', async () => {
    await asar.createPackageWithOptions('test/input/packthis/', 'tmp/packthis-without-hidden-api.asar', { dot: false })
    return compFiles('tmp/packthis-without-hidden-api.asar', 'test/expected/packthis-without-hidden.asar')
  })
  it('should create archive from directory (with transformed files)', async () => {
    await asar.createPackageWithOptions('test/input/packthis/', 'tmp/packthis-api-transformed.asar', { transform })
    return compFiles('tmp/packthis-api-transformed.asar', 'test/expected/packthis-transformed.asar')
  })
  it('should create archive from directory (with nothing packed)', async () => {
    await asar.createPackageWithOptions('test/input/packthis/', 'tmp/packthis-api-unpacked.asar', { unpackDir: '**' })
    await compFiles('tmp/packthis-api-unpacked.asar', 'test/expected/packthis-all-unpacked.asar')
    return compDirs('tmp/packthis-api-unpacked.asar.unpacked', 'test/expected/extractthis')
  })
  it('should list files/dirs in archive', async () => {
    return assertPackageListEquals(asar.listPackage('test/input/extractthis.asar'), 'test/expected/extractthis-filelist.txt')
  })
  it('should list files/dirs in archive with option', async () => {
    return assertPackageListEquals(asar.listPackage('test/input/extractthis-unpack-dir.asar', { isPack: true }), 'test/expected/extractthis-filelist-with-option.txt')
  })
  it('should extract a text file from archive', async () => {
    const actual = asar.extractFile('test/input/extractthis.asar', 'dir1/file1.txt').toString('utf8')
    const expected = await fs.readFile('test/expected/extractthis/dir1/file1.txt', 'utf8')
    return compFileLists(actual, expected)
  })
  it('should extract a binary file from archive', async () => {
    const actual = asar.extractFile('test/input/extractthis.asar', 'dir2/file2.png')
    const expected = await fs.readFile('test/expected/extractthis/dir2/file2.png')
    return assert.strictEqual(actual.toString(), expected.toString())
  })
  it('should extract a binary file from archive with unpacked files', async () => {
    const actual = asar.extractFile('test/input/extractthis-unpack.asar', 'dir2/file2.png')
    const expected = await fs.readFile('test/expected/extractthis/dir2/file2.png')
    return assert.strictEqual(actual.toString(), expected.toString())
  })
  it('should extract an archive', async () => {
    asar.extractAll('test/input/extractthis.asar', 'tmp/extractthis-api/')
    return compDirs('tmp/extractthis-api/', 'test/expected/extractthis')
  })
  it('should extract an archive with unpacked files', async () => {
    asar.extractAll('test/input/extractthis-unpack.asar', 'tmp/extractthis-unpack-api/')
    return compDirs('tmp/extractthis-unpack-api/', 'test/expected/extractthis')
  })
  it('should extract a binary file from archive with unpacked files', async () => {
    const actual = asar.extractFile('test/input/extractthis-unpack-dir.asar', 'dir1/file1.txt')
    const expected = await fs.readFile('test/expected/extractthis/dir1/file1.txt')
    assert.strictEqual(actual.toString(), expected.toString())
  })
  it('should extract an archive with unpacked dirs', async () => {
    asar.extractAll('test/input/extractthis-unpack-dir.asar', 'tmp/extractthis-unpack-dir-api/')
    return compDirs('tmp/extractthis-unpack-dir-api/', 'test/expected/extractthis')
  })
  it('should handle multibyte characters in paths', async () => {
    await asar.createPackageWithOptions('test/input/packthis-unicode-path/', 'tmp/packthis-unicode-path.asar', {
      globOptions: {
        nosort: true
      }
    })
    return compFiles('tmp/packthis-unicode-path.asar', 'test/expected/packthis-unicode-path.asar')
  })
  it('should extract a text file from archive with multibyte characters in path', async () => {
    const actual = asar.extractFile('test/expected/packthis-unicode-path.asar', 'dir1/女の子.txt').toString('utf8')
    const expected = await fs.readFile('test/input/packthis-unicode-path/dir1/女の子.txt', 'utf8')
    return compFileLists(actual, expected)
  })
  it('should create files/directories whose names are properties of Object.prototype', async () => {
    await asar.createPackage('test/input/packthis-object-prototype/', 'tmp/packthis-object-prototype.asar')
    return compFiles('tmp/packthis-object-prototype.asar', 'test/expected/packthis-object-prototype.asar')
  })
  it('should extract files/directories whose names are properties of Object.prototype', () => {
    asar.extractAll('test/expected/packthis-object-prototype.asar', 'tmp/packthis-object-prototype/')
    return compDirs('test/input/packthis-object-prototype/', 'tmp/packthis-object-prototype')
  })
})
