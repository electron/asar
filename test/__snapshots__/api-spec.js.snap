// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api should create archive from directory (with nothing packed) 1`] = `
Object {
  "files": Object {
    ".hiddenfile.txt": Object {
      "integrity": Object {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": Array [
          "d400d9a74f67e724a4793173d73e1403bd6b734930632f4eee50a3a5a1e5a478",
        ],
        "hash": "d400d9a74f67e724a4793173d73e1403bd6b734930632f4eee50a3a5a1e5a478",
      },
      "size": 19,
      "unpacked": true,
    },
    "dir1": Object {
      "files": Object {
        "file1.txt": Object {
          "integrity": Object {
            "algorithm": "SHA256",
            "blockSize": 4194304,
            "blocks": Array [
              "420149d3f852894ba7f32e9d3ec7d8919ee2451724bf2580b0186cd373bd6d82",
            ],
            "hash": "420149d3f852894ba7f32e9d3ec7d8919ee2451724bf2580b0186cd373bd6d82",
          },
          "size": 9,
          "unpacked": true,
        },
      },
      "unpacked": true,
    },
    "dir2": Object {
      "files": Object {
        "file2.png": Object {
          "integrity": Object {
            "algorithm": "SHA256",
            "blockSize": 4194304,
            "blocks": Array [
              "cc402b796dc92b2b1f3a6d09515003d8400e63d8acaffc967e49c0cf015fcffe",
            ],
            "hash": "cc402b796dc92b2b1f3a6d09515003d8400e63d8acaffc967e49c0cf015fcffe",
          },
          "size": 182,
          "unpacked": true,
        },
        "file3.txt": Object {
          "integrity": Object {
            "algorithm": "SHA256",
            "blockSize": 4194304,
            "blocks": Array [
              "a665a45920422f9d417e4867efdc4fb8a04a1f3fff1fa07e998e86f7f7a27ae3",
            ],
            "hash": "a665a45920422f9d417e4867efdc4fb8a04a1f3fff1fa07e998e86f7f7a27ae3",
          },
          "size": 3,
          "unpacked": true,
        },
      },
      "unpacked": true,
    },
    "emptyfile.txt": Object {
      "integrity": Object {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": Array [
          "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
        ],
        "hash": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
      },
      "size": 0,
      "unpacked": true,
    },
    "file0.txt": Object {
      "integrity": Object {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": Array [
          "41a978be88ff87a57308bf234106a57ebf9aa0971e2101e75113547a815be72b",
        ],
        "hash": "41a978be88ff87a57308bf234106a57ebf9aa0971e2101e75113547a815be72b",
      },
      "size": 13,
      "unpacked": true,
    },
  },
}
`;

exports[`api should create archive from directory (with nothing packed) 2`] = `
Array [
  Object {
    "content": "file one.",
    "name": "dir1/file1.txt",
  },
  "dir2/file2.png",
  Object {
    "content": "123",
    "name": "dir2/file3.txt",
  },
  "dir1",
  "dir2",
  Object {
    "content": "This file is hidden",
    "name": ".hiddenfile.txt",
  },
  Object {
    "content": "",
    "name": "emptyfile.txt",
  },
  Object {
    "content": "file0 content",
    "name": "file0.txt",
  },
]
`;

exports[`api should create archive from directory (with transformed files) 1`] = `
Object {
  "files": Object {
    ".hiddenfile.txt": Object {
      "integrity": Object {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": Array [
          "d400d9a74f67e724a4793173d73e1403bd6b734930632f4eee50a3a5a1e5a478",
        ],
        "hash": "d400d9a74f67e724a4793173d73e1403bd6b734930632f4eee50a3a5a1e5a478",
      },
      "offset": "0",
      "size": 19,
    },
    "dir1": Object {
      "files": Object {
        "file1.txt": Object {
          "integrity": Object {
            "algorithm": "SHA256",
            "blockSize": 4194304,
            "blocks": Array [
              "420149d3f852894ba7f32e9d3ec7d8919ee2451724bf2580b0186cd373bd6d82",
            ],
            "hash": "420149d3f852894ba7f32e9d3ec7d8919ee2451724bf2580b0186cd373bd6d82",
          },
          "offset": "19",
          "size": 9,
        },
      },
    },
    "dir2": Object {
      "files": Object {
        "file2.png": Object {
          "integrity": Object {
            "algorithm": "SHA256",
            "blockSize": 4194304,
            "blocks": Array [
              "cc402b796dc92b2b1f3a6d09515003d8400e63d8acaffc967e49c0cf015fcffe",
            ],
            "hash": "cc402b796dc92b2b1f3a6d09515003d8400e63d8acaffc967e49c0cf015fcffe",
          },
          "offset": "28",
          "size": 182,
        },
        "file3.txt": Object {
          "integrity": Object {
            "algorithm": "SHA256",
            "blockSize": 4194304,
            "blocks": Array [
              "a665a45920422f9d417e4867efdc4fb8a04a1f3fff1fa07e998e86f7f7a27ae3",
            ],
            "hash": "a665a45920422f9d417e4867efdc4fb8a04a1f3fff1fa07e998e86f7f7a27ae3",
          },
          "offset": "210",
          "size": 3,
        },
      },
    },
    "emptyfile.txt": Object {
      "integrity": Object {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": Array [
          "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
        ],
        "hash": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
      },
      "offset": "213",
      "size": 0,
    },
    "file0.txt": Object {
      "integrity": Object {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": Array [
          "41a978be88ff87a57308bf234106a57ebf9aa0971e2101e75113547a815be72b",
        ],
        "hash": "41a978be88ff87a57308bf234106a57ebf9aa0971e2101e75113547a815be72b",
      },
      "offset": "213",
      "size": 13,
    },
  },
}
`;

exports[`api should create archive from directory (without hidden files) 1`] = `
Object {
  "files": Object {
    "dir1": Object {
      "files": Object {
        "file1.txt": Object {
          "integrity": Object {
            "algorithm": "SHA256",
            "blockSize": 4194304,
            "blocks": Array [
              "420149d3f852894ba7f32e9d3ec7d8919ee2451724bf2580b0186cd373bd6d82",
            ],
            "hash": "420149d3f852894ba7f32e9d3ec7d8919ee2451724bf2580b0186cd373bd6d82",
          },
          "offset": "0",
          "size": 9,
        },
      },
    },
    "dir2": Object {
      "files": Object {
        "file2.png": Object {
          "integrity": Object {
            "algorithm": "SHA256",
            "blockSize": 4194304,
            "blocks": Array [
              "cc402b796dc92b2b1f3a6d09515003d8400e63d8acaffc967e49c0cf015fcffe",
            ],
            "hash": "cc402b796dc92b2b1f3a6d09515003d8400e63d8acaffc967e49c0cf015fcffe",
          },
          "offset": "9",
          "size": 182,
        },
        "file3.txt": Object {
          "integrity": Object {
            "algorithm": "SHA256",
            "blockSize": 4194304,
            "blocks": Array [
              "a665a45920422f9d417e4867efdc4fb8a04a1f3fff1fa07e998e86f7f7a27ae3",
            ],
            "hash": "a665a45920422f9d417e4867efdc4fb8a04a1f3fff1fa07e998e86f7f7a27ae3",
          },
          "offset": "191",
          "size": 3,
        },
      },
    },
    "emptyfile.txt": Object {
      "integrity": Object {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": Array [
          "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
        ],
        "hash": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
      },
      "offset": "194",
      "size": 0,
    },
    "file0.txt": Object {
      "integrity": Object {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": Array [
          "41a978be88ff87a57308bf234106a57ebf9aa0971e2101e75113547a815be72b",
        ],
        "hash": "41a978be88ff87a57308bf234106a57ebf9aa0971e2101e75113547a815be72b",
      },
      "offset": "194",
      "size": 13,
    },
  },
}
`;

exports[`api should create archive from directory 1`] = `
Object {
  "files": Object {
    ".hiddenfile.txt": Object {
      "integrity": Object {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": Array [
          "d400d9a74f67e724a4793173d73e1403bd6b734930632f4eee50a3a5a1e5a478",
        ],
        "hash": "d400d9a74f67e724a4793173d73e1403bd6b734930632f4eee50a3a5a1e5a478",
      },
      "offset": "0",
      "size": 19,
    },
    "dir1": Object {
      "files": Object {
        "file1.txt": Object {
          "integrity": Object {
            "algorithm": "SHA256",
            "blockSize": 4194304,
            "blocks": Array [
              "420149d3f852894ba7f32e9d3ec7d8919ee2451724bf2580b0186cd373bd6d82",
            ],
            "hash": "420149d3f852894ba7f32e9d3ec7d8919ee2451724bf2580b0186cd373bd6d82",
          },
          "offset": "19",
          "size": 9,
        },
      },
    },
    "dir2": Object {
      "files": Object {
        "file2.png": Object {
          "integrity": Object {
            "algorithm": "SHA256",
            "blockSize": 4194304,
            "blocks": Array [
              "cc402b796dc92b2b1f3a6d09515003d8400e63d8acaffc967e49c0cf015fcffe",
            ],
            "hash": "cc402b796dc92b2b1f3a6d09515003d8400e63d8acaffc967e49c0cf015fcffe",
          },
          "offset": "28",
          "size": 182,
        },
        "file3.txt": Object {
          "integrity": Object {
            "algorithm": "SHA256",
            "blockSize": 4194304,
            "blocks": Array [
              "a665a45920422f9d417e4867efdc4fb8a04a1f3fff1fa07e998e86f7f7a27ae3",
            ],
            "hash": "a665a45920422f9d417e4867efdc4fb8a04a1f3fff1fa07e998e86f7f7a27ae3",
          },
          "offset": "210",
          "size": 3,
        },
      },
    },
    "emptyfile.txt": Object {
      "integrity": Object {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": Array [
          "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
        ],
        "hash": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
      },
      "offset": "213",
      "size": 0,
    },
    "file0.txt": Object {
      "integrity": Object {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": Array [
          "41a978be88ff87a57308bf234106a57ebf9aa0971e2101e75113547a815be72b",
        ],
        "hash": "41a978be88ff87a57308bf234106a57ebf9aa0971e2101e75113547a815be72b",
      },
      "offset": "213",
      "size": 13,
    },
  },
}
`;

exports[`api should create package from array of NodeJS.ReadableStreams 1`] = `
Object {
  "files": Object {
    "x1": Object {
      "files": Object {
        "file1.txt": Object {
          "integrity": Object {
            "algorithm": "SHA256",
            "blockSize": 4194304,
            "blocks": Array [
              "be4f0d7f4574cda8a2c7b45495a252df4ed0ceaf8eedb692fb00e3409444b591",
            ],
            "hash": "be4f0d7f4574cda8a2c7b45495a252df4ed0ceaf8eedb692fb00e3409444b591",
          },
          "size": 8,
          "unpacked": true,
        },
      },
      "unpacked": true,
    },
    "x2": Object {
      "files": Object {
        "file2.txt": Object {
          "integrity": Object {
            "algorithm": "SHA256",
            "blockSize": 4194304,
            "blocks": Array [
              "586ea213ecd662832d578a6d53b84f34a44df8a5d23555215ec11dbd708dfdb6",
            ],
            "hash": "586ea213ecd662832d578a6d53b84f34a44df8a5d23555215ec11dbd708dfdb6",
          },
          "offset": "0",
          "size": 8,
        },
      },
    },
    "y3": Object {
      "files": Object {
        "file3.txt": Object {
          "integrity": Object {
            "algorithm": "SHA256",
            "blockSize": 4194304,
            "blocks": Array [
              "33e658c933a8d7ac2706b5ea44d0887c9a6617d33c89cfd7aa8a8a5043c5577e",
            ],
            "hash": "33e658c933a8d7ac2706b5ea44d0887c9a6617d33c89cfd7aa8a8a5043c5577e",
          },
          "offset": "17",
          "size": 10,
        },
        "x1": Object {
          "files": Object {
            "file4.txt": Object {
              "integrity": Object {
                "algorithm": "SHA256",
                "blockSize": 4194304,
                "blocks": Array [
                  "dbeebbfbd66e243dd5c2e2fd96d83318269c89d99ebb0aa1a5a1f027092db5e1",
                ],
                "hash": "dbeebbfbd66e243dd5c2e2fd96d83318269c89d99ebb0aa1a5a1f027092db5e1",
              },
              "size": 9,
              "unpacked": true,
            },
          },
          "unpacked": true,
        },
        "z1": Object {
          "files": Object {
            "x2": Object {
              "files": Object {
                "file5.txt": Object {
                  "integrity": Object {
                    "algorithm": "SHA256",
                    "blockSize": 4194304,
                    "blocks": Array [
                      "74abc479d95c489f300f18086bcab63f188b89f6d309725ea613cf6a8d487a88",
                    ],
                    "hash": "74abc479d95c489f300f18086bcab63f188b89f6d309725ea613cf6a8d487a88",
                  },
                  "offset": "8",
                  "size": 9,
                },
              },
            },
          },
        },
      },
    },
    "z4": Object {
      "files": Object {
        "w1": Object {
          "files": Object {
            "file6.txt": Object {
              "integrity": Object {
                "algorithm": "SHA256",
                "blockSize": 4194304,
                "blocks": Array [
                  "10ad5315269b2145a0990eeb305f792c522f5d8efda2c58e180fbfed5be231cd",
                ],
                "hash": "10ad5315269b2145a0990eeb305f792c522f5d8efda2c58e180fbfed5be231cd",
              },
              "offset": "27",
              "size": 8,
            },
          },
        },
      },
    },
  },
}
`;

exports[`api should create package from array of NodeJS.ReadableStreams 2`] = `
Array [
  Object {
    "content": "fileone
",
    "name": "x1/file1.txt",
  },
  Object {
    "content": "filefour
",
    "name": "y3/x1/file4.txt",
  },
  "y3/x1",
  "x1",
  "y3",
]
`;

exports[`api should handle multibyte characters in paths 1`] = `
Object {
  "files": Object {
    "dir1": Object {
      "files": Object {
        "file1.txt": Object {
          "integrity": Object {
            "algorithm": "SHA256",
            "blockSize": 4194304,
            "blocks": Array [
              "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
            ],
            "hash": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
          },
          "offset": "0",
          "size": 0,
        },
        "女の子.txt": Object {
          "integrity": Object {
            "algorithm": "SHA256",
            "blockSize": 4194304,
            "blocks": Array [
              "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
            ],
            "hash": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
          },
          "offset": "0",
          "size": 0,
        },
      },
    },
  },
}
`;
